src/dark_mind.c

```diff
@@ -0,0 +1,2218 @@
+    }
+    if (r7->unk0.base.unk1 == 1)
+        r7->unk0.base.flags |= 8;
+}
+
+void sub_08101560(struct DarkMind *r7)
+{
+    struct Object4 *r3;
+
+    if (r7->unk0.unk9F == 32 || r7->unk0.unk9F == 64
+        || r7->unk0.unk9F == 94 || r7->unk0.unk9F == 124
+        || r7->unk0.unk9F == 148 || r7->unk0.unk9F == 168
+        || r7->unk0.unk9F == 186 || r7->unk0.unk9F == 202
+        || r7->unk0.unk9F == 216 || r7->unk0.unk9F == 230
+        || r7->unk0.unk9F == 240 || r7->unk0.unk9F == 248
+        || r7->unk0.unk9F == 255)
+        Macro_081013C8(r7, r3);
+    ++r7->unk0.unk9F;
+    if (++r7->unk0.base.counter > 248)
+    {
+        sub_080700D8(&r7->unk0);
+        r7->unk0.base.flags |= 0x1000;
+    }
+}
+
+void sub_08101630(struct DarkMind *r5)
+{
+    if (r5->unk0.unk80 <= 0)
+        sub_08101350(r5);
+    else
+    {
+        r5->unkD5 = 0;
+        DarkMindSetFunc(r5, 0, sub_08101784);
+        r5->unk0.base.xspeed = 0;
+        r5->unk0.base.yspeed = 0;
+        r5->unk0.base.flags |= 0x40;
+        r5->unk0.kirby3 = sub_0803D368(&r5->unk0);
+        if (r5->unk0.base.x > r5->unk0.kirby3->base.base.x)
+            r5->unk0.base.flags |= 1;
+        else
+            r5->unk0.base.flags &= ~1;
+        r5->unk0.unkA2 = r5->unk0.base.y >> 8;
+        switch (FindIdx(Rand16()))
+        {
+        case 0:
+        case 1:
+            r5->unk0.unk9F = 1;
+            break;
+        case 2:
+        case 3:
+            r5->unk0.unk9F = 2;
+            break;
+        case 4:
+        default:
+            r5->unk0.unk9F = 3;
+            break;
+        }
+        sub_0803E2B0(&r5->unk0.base, -28, -24, 28, 18);
+        r5->unk0.unk91 = -8;
+        r5->unk0.unk92 = -8;
+        r5->unk0.unk93 = -8;
+        r5->unk0.unk94 = -8;
+        r5->unk0.unk95 = -8;
+        r5->unk0.unk96 = -8;
+        r5->unk0.unk97 = -32;
+    }
+}
+
+void sub_08101784(struct DarkMind *r3)
+{
+    r3->unk0.base.flags |= 4;
+    r3->unk0.base.yspeed = gUnk_08357256[(r3->unk0.base.counter >> 4) & 3];
+    ++r3->unk0.base.counter;
+    if (r3->unk0.unkA2 == (s16)(r3->unk0.base.y >> 8))
+    {
+        if (!r3->unk0.unk9F)
+        {
+            r3->unk0.base.yspeed = 0;
+            switch (r3->unk0.subtype)
+            {
+            case 0:
+                sub_081007A8(r3);
+                break;
+            case 1:
+                sub_08100858(r3);
+                break;
+            case 2:
+                sub_081009A4(r3);
+                break;
+            default:
+                sub_08100BD0(r3);
+                break;
+            }
+        }
+        else
+        {
+            --r3->unk0.unk9F;
+        }
+    }
+}
+
+void sub_0810181C(struct DarkMind *r4)
+{
+    struct DarkMind *r5 = r4;
+
+    DarkMindSetFunc(r4, 0, sub_08101B28);
+    r4->unk0.base.xspeed = 0;
+    r4->unk0.base.yspeed = 0;
+    r4->unk0.base.flags |= 0x40;
+    r4->unk0.kirby3 = sub_0803D46C(&r4->unk0);
+    if (r4->unk0.base.x > r4->unk0.kirby3->base.base.x)
```

Comment:
This pattern is also pretty frequent. Maybe we should create a macro for this (ObjChooseXDir)?